plugins/inputs/sqlserver/connectionstring.go

```
@@ -0,0 +1,121 @@
+	if len(connectionString) == 0 {
+		return
+	}
+
+	trimmedConnectionString := strings.TrimSpace(connectionString)
+
+	if strings.HasPrefix(trimmedConnectionString, odbcPrefix) {
+		i := strings.Index(trimmedConnectionString, odbcPrefix)
+		connectionStringWithoutOdbc := trimmedConnectionString[i+len(odbcPrefix):]
+		sqlInstance, databaseName = parseConnectionStringKeyValue(connectionStringWithoutOdbc)
+	} else if strings.HasPrefix(trimmedConnectionString, "sqlserver://") {
+		sqlInstance, databaseName = parseConnectionStringURL(trimmedConnectionString)
+	} else {
+		sqlInstance, databaseName = parseConnectionStringKeyValue(trimmedConnectionString)
+	}
+
+	return
+}
+
+// parseConnectionStringKeyValue parses a "key=value;" connection string and returns the SQL instance and database name
+func parseConnectionStringKeyValue(connectionString string) (sqlInstance string, databaseName string) {
+	sqlInstance = emptyServerName
+	databaseName = emptyDatabaseName
+
+	keyValuePairs := strings.Split(connectionString, ";")
+	for _, keyValuePair := range keyValuePairs {
+		if len(keyValuePair) == 0 {
+			continue
+		}
+
+		keyAndValue := strings.SplitN(keyValuePair, "=", 2)
+		key := strings.TrimSpace(strings.ToLower(keyAndValue[0]))
+		if len(key) == 0 {
+			continue
+		}
+
+		var value string = ""
+		if len(keyAndValue) > 1 {
+			value = strings.TrimSpace(keyAndValue[1])
+		}
+
+		if isInstanceIdentifier(key) {
+			sqlInstance = value
+			if databaseName != emptyDatabaseName {
+				break
+			}
+		} else if isDatabaseIdentifier(key) {
+			databaseName = value
+			if sqlInstance != emptyServerName {
+				break
+			}
+		}
+	}
+
+	return
+}
+
+// parseConnectionStringURL parses a URL-formatted connection string and returns the SQL instance and database name
+func parseConnectionStringURL(connectionString string) (sqlInstance string, databaseName string) {
+	sqlInstance = emptyServerName
+	databaseName = emptyDatabaseName
+
+	u, err := url.Parse(connectionString)
+	if err != nil {
```

Do we need to log the error if unable to parse?

---

I actually thought about this for a while when I originally wrote it. I decided not to log to avoid exposing any secrets possibly contained in the connection string. If we fail to parse here, then the SQL connection will certainly fail, so it won't be failing silently.